### PR TITLE
Verify that the configuration keys in the file are valid.

### LINKED ISSUE #19515: Improve validation of daemon config
Per https://github.com/docker/docker/pull/19457#issuecomment-173329801

>  invalid config options are currently silently ignored.
> This could be really bad particularly with `tls-verify`, if someone does something like `tlsverify` instead.

Not critical, but It would be nice to improve this.

I'm adding this to the 1.10 milestone for if someone wants to work on it, but I'm fine with shipping 1.10 without this implemented.

### PR DESCRIPTION
**Pull request by calavera**
Return an error if any of the keys don't match valid flags.
Flatten configuration file by embedding TLSOptions and LogConfig.

Fixes #19515 

/cc @cpuguy83

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by tiborvass** [comment]
+1

**Comment by thaJeztah** [comment]
@calavera closes https://github.com/docker/docker/issues/19515 I think?

**Comment by thaJeztah** [comment]
+1

**Comment by calavera** [comment]
> @calavera closes #19515 I think?

yep

**Comment by cpuguy83** [comment]
Error message isn't right (in general), also seems to be flagging `hosts` as not a valid key?

```
$ docker daemon
unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: hosts: unknown configuration key
$
```

**Comment by calavera** [comment]
> also seems to be flagging hosts as not a valid key?

that's very interesting, I wonder if there is another bug hidden in that validation. Checking.

**Comment by calavera** [comment]
So, it turned out that there were more hidden issues behind all this :dizzy_face: 

1. The `tls-verify` configuration key was wrong, the flag is named `-tlsverify`, therefore the key should be `tlsverify`.
2. The `mflag.Merge` function had a bug where it raised a panic if any of the flag sets didn't have any flag.
3. That same function used a wrapper around merged flags that avoided the `host` flag to be recognized as a named flag and it was being ignored in the conflicts detection.

All that should be fixed now

**Comment by calavera** [comment]
This is the message when a key in the configuration file is not valid:

```
unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives don't match any configuration option: foo
```

File: daemon/config.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ type CommonConfig struct {
 	Hosts      []string         `json:"hosts,omitempty"`
 	LogLevel   string           `json:"log-level,omitempty"`
 	TLS        bool             `json:"tls,omitempty"`
-	TLSVerify  bool             `json:"tls-verify,omitempty"`
+	TLSVerify  bool             `json:"tlsverify,omitempty"`
 	TLSOptions CommonTLSOptions `json:"tls-opts,omitempty"`
```
Comment (tiborvass) [inline daemon/config.go]: there is no tls-opts flag at all :(

Comment (tiborvass) [inline daemon/config.go]: `--tlscacert`, `--tlscert`, `--tlskey` are the 3 flags

Comment (calavera) [inline daemon/config.go]: they are defined in the `CommonTLSOptions` struct. The conflict logic checks those fields when it finds inner structs like this one. The keys for the options are still called `tlscacert`, `tlscert` and `tlskey`.

File: daemon/config.go [hunk 2]
```diff
@@ -215,16 +215,43 @@ func configValuesSet(config map[string]interface{}) map[string]interface{} {
...
+	// 1. Search keys from the file that we don't recognize as flags.
+	unknownKeys := make(map[string]interface{})
+	for key, value := range config {
+		flagName := "-" + key
+		if flag := flags.Lookup(flagName); flag == nil {
+			unknownKeys[key] = value
+		}
+	}
+
+	// 2. Discard keys that might have a given name, like `labels`.
```
Comment (cpuguy83) [inline daemon/config.go]: What does this comment mean?

**Comment by cpuguy83** [comment]
Other than a question on the code comment, LGTM.

**Comment by tiborvass** [comment]
I would like to see `TLSOptions` be embedded so that we can remove `tls-opts` which is not a real flag.

**Comment by calavera** [comment]
@tiborvass I added a second commit with that change so it can be reviewed in isolation, I don't think that belongs to this PR, but we can merge them together, I guess.

**Comment by tiborvass** [comment]
@calavera thank you!

LGTM

**Comment by estesp** [comment]
LGTM

**Pull request merged**